src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java

```diff
@@ -114,24 +116,39 @@ public void close() {
      * @param type  action type.  Either "app_data.wsConnAdded" or "app_data.wsConnRemoved"
      * @param channelID
      */
-    private void updateClientConnections(String type, String channelID) {
+    private void updateClientConnections(String type, String channelID, String userKey) {
         List<ServerEventQueue> conns = ServerEventManager.getEvQueueList();
-        String baseChannel = channelID.replace("__viewer", "");
-        String viewerChannel = baseChannel + "__viewer";
-        ArrayList<String> baseList = new ArrayList<>();
-        ArrayList<String> viewerList = new ArrayList<>();
-        FluxAction action = new FluxAction(type);
-        for (ServerEventQueue q : conns) {
-            if (q.getChannel().equals(baseChannel)) {
-                baseList.add(q.getConnID());
-            }else if (q.getChannel().equals(viewerChannel)) {
-                viewerList.add(q.getConnID());
+        for (ServerEventQueue seq : conns) {
+            // need to notify clients that are affected by update
+            if (seq.getChannel().equals(channelID) || seq.getUserKey().equals(userKey)) {
+                // Creates a map of all the channels and its connections that is visible to this user.
+                // This user has knowledge of all connections started by this user, as well as connections associated with this user's channel.
+                Map<String, List<String>> connInfo = new HashMap<>();
+                conns.stream()
+                        .filter(q -> q.getChannel().equals(channelID) || q.getUserKey().equals(userKey))
+                        .forEach(q -> {
+                            List<String> l = connInfo.get(q.getChannel());
+                            if (l == null) {
+                                l = new ArrayList<>();
+                                connInfo.put(q.getChannel(), l);
+                            }
+                            if (!l.contains(q.getConnID())) l.add(q.getConnID());
+                        });
+                FluxAction action = new FluxAction(type, connInfo);
+                ServerEventManager.fireAction(action, new ServerEvent.EventTarget(ServerEvent.Scope.SELF, seq.getConnID(), null, null));
```

Comment:
This is done to support JS external viewer and python API(?).  I am guessing this information may be useful in the future as well.